pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,306 @@
+// the traffic through tunnels between Gateway and other Nodes, as otherwise the
+// traffic will not go through tunnels in those modes.
+type MCPodRouteController struct {
+	mcClient        mcclientset.Interface
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+	ofClient        openflow.Client
+	interfaceStore  interfacestore.InterfaceStore
+	nodeConfig      *config.NodeConfig
+	queue           workqueue.RateLimitingInterface
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	gwInformer      cache.SharedIndexInformer
+	gwLister        mclisters.GatewayLister
+	activeGWName    string
+	namespace       string
+	mutex           sync.Mutex
+	installedPods   map[string]string
+}
+
+type endpointChangeInfo struct {
+	podIP     string
+	podNodeIP string
+}
+
+func NewMCPodRouteController(
+	mcClient mcclientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteController"),
+		podInformer:     podInformer,
+		podLister:       corelisters.NewPodLister(podInformer.GetIndexer()),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		installedPods:   make(map[string]string),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+
+	controller.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// Gateway webhook will guarantee that at most one Gateway in a cluster.
+	if isDelete {
```

Comment:
Question - for Gateway change, will we always get a "Gateway" delete first?

Reply:
yes, new Gateway won't be created if there is already a Gateway exists. The Gateway webhook will check this.

Reply:
In my mind, we need not to know it is a delete or not. For any Gateway even, just get the Gateway in the Namespace and check whether any exists and matches the current Node. It is simpler and more robust than assuming particular event order or webhook logic.

Reply:
Done